aeneas/src/ir/VariantSolver.v3

```diff
@@ -83,6 +101,197 @@ def ON_STACK = -1;
 def EMPTY_FIELD = VariantField.new(null, []);
 def NO_FIELDS = Array<VariantField>.new(0);
 
+class VariantSolver(vnorm: VariantNormalizer, usePacking: bool, useTagScalar: bool) {
+	var normFields: Array<Array<Type>>;
+
+	var fields: Vector<CaseField>; // an ordering of fields to solve for during packing
+
+	// final solutions
+	var scalarAsns: HashMap<CaseField, int>; // field location -> scalar idx
+	var intervalAsns: HashMap<CaseField, Interval>; // field location -> interval within scalar
+	var state: Array<VariantPattern>;
+
+	var explicitTagLength: byte;
+	var explicitTag: (int, Interval) = (-1, Interval(0, 0));
+
+	var curRep: Vector<Scalar.set>;
+	var curUsed: Vector<byte>;
+	
+	// Perform recursive backtracking on the potential representation.
+	// With each scalar in each case, we have several choices to make:
+	// 1. Pack this scalar with an existing scalar
+	// 2. Append this scalar as a new scalar.
+	// We can only determine distinguishability after all the scalars have been assigned. Heuristics will help speed this part up.
+	private def tryRepresentationForField(curCase: int, curField: int) -> bool {
+		if (curCase >= normFields.length) {
+			// XXX: check distinguishable
+			return solvePacking();
+		}
+		if (curField >= normFields[curCase].length) {
+			// move on to next case
+			curUsed = Vector<byte>.new();
+			for (i < curRep.length) curUsed.put(0);
+			return tryRepresentationForField(curCase + 1, 0);
+		}
+		
+		var sc = vnorm.nc.GetScalar(vnorm.rn.ra.compiler, vnorm.rn.ra.prog, normFields[curCase][curField]);
```

Comment:
The `VariantNormalizer` should probably have a method that hides its configuration and the compiler and program arguments.